Bug-1956484 WebAuthn support in web extensions (#43653)
mdn/content: files/en-us/mozilla/add-ons/webextensions/use_the_web_authn_api/index.md

You can extract the origin from a credential response on any browser like this:

```js
let clientData = JSON.parse(
  new TextDecoder().decode(publicKeyCredential.response.clientDataJSON),
);
console.log(clientData.origin);
// Chrome:  chrome-extension://mabekielmoibbmlepeohhncklpnjmcpk
// Firefox: moz-extension://ngpncaopklanhjklijieoihgbhbgknjjdklmlpagjoaobbpmknfgmhgghbadgoai
```

## Setting up WebAuthn in a web extension

This guide walks through building an extension that creates and retrieves WebAuthn credentials. It does this by enabling your extension to supply the JSON that defines the RP ID from a domain the extension has host permission for.

### Configure the manifest

In your extension's `manifest.json` file, declare `host_permissions` for the domain or domains whose RP ID you want to use. This setting grants the extension permission to run content scripts on those domains, which is the prerequisite for asserting the domain as an RP ID.

```json
{
  "manifest_version": 3,
  "name": "WebAuthn Extension",
  "version": "1.0",
  "description": "Registers and authenticates WebAuthn credentials.",
  "action": {
    "default_popup": "popup.html"
  },

>   "permissions": ["storage"],

Reviewer: ```suggestion
```

  "host_permissions": ["https://*/*"]
}
```

> [!NOTE]
> The broad `"https://*/*"` pattern grants access to all HTTPS domains. For your extension, set this to the domains it needs (e.g., `"https://example.com/*"`).

> 

Reviewer: Might also be worth adding a warning: the example here is written for use in extension popups, but the flow does not actually work because the popup is closed as soon as the prompt for credentials appears. A work-around is to open the page in a new tab. Relevant bug is https://bugzilla.mozilla.org/show_bug.cgi?id=2026687

Author: Done

### Collect WebAuthn options

Add a mechanism to enter the registration JSON. In this case, a pop-up; you could also use an extension page. This example uses a simple `popup.html` with a `textarea` for the JSON input and two buttons: one for registration and one for authentication.

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <title>WebAuthn Extension</title>
  </head>
  <body>
    <h1>WebAuthn</h1>
    <textarea id="optionsText" placeholder="Enter options JSON here"></textarea>
    <button id="registerButton">Register (navigator.credentials.create)</button>
    <button id="authButton">Authenticate (navigator.credentials.get)</button>
    <script src="popup.js"></script>
  </body>
</html>
```

### Register the credentials

> In your extension script, parse the options JSON and call `navigator.credentials.create()`. The key detail is that you set `rp.id` to a domain covered by your extension's host permissions, even though your extension is not running on that domain.

Reviewer: Explicitly mentioning the `publicKey` property and its type because it is not obvious from the example.

```suggestion
In your extension script, parse the options JSON and call `navigator.credentials.create()` with `publicKey` in the {{domxref("PublicKeyCredential")}} format. The key detail is that you set `rp.id` to a domain covered by your extension's host permissions, even though your extension is not running on that domain.
```

> Binary fields, such as `challenge` and `user.id`, must be converted to `ArrayBuffer` instances. If your JSON input uses Base64 strings, add code to convert them:

Reviewer: ```suggestion
Binary fields, such as `challenge` and `user.id`, must be passed as typed arrays or `ArrayBuffer` instances. If your JSON input uses base64-encoded strings, the {{jsxref("Uint8Array.fromBase64")}} method can be used to convert them.
```

Correction: not just ArrayBuffer, typed arrays such as Uint8Array are acceptable too. I suggested an edit including a reference to the relevant documentation, not sure if the macro syntax is fully correct, let's double-check that after updating the PR.

Also pointing to the fromBase64 method to avoid the need for the custom helper, simplifying the code example. Please update the example below.

```js
function base64ToArrayBuffer(base64) {
  const binaryString = window.atob(base64);
  const bytes = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  return bytes.buffer;
}

function convertOptions(options) {
  if (typeof options.challenge === "string") {
    options.challenge = base64ToArrayBuffer(options.challenge);
  }
  if (options.user && typeof options.user.id === "string") {
    options.user.id = base64ToArrayBuffer(options.user.id);
  }
  if (options.allowCredentials) {
    options.allowCredentials = options.allowCredentials.map((cred) => ({
      ...cred,
      id: base64ToArrayBuffer(cred.id),
    }));
  }
  return options;
}
```

> 

Reviewer: Drop this whole code block:
- `base64ToArrayBuffer` is redundant to the availability of `Uint8Array.fromBase64()`
- Drop `convertOptions` because the method misleadingly suggests that there many options common to the `get()` and `create()` methods, and that some parts may be optional, when that is not the case. Only the `challenge` property is common to the two, and that property, as well as `user.id`, are required in `navigator.credentials.create`. On the other hand, the `allowCredentials` field is not recognized by `create()`, only by `navigator.credentials.get()`, and it is also optional.

```suggestion
```

You can then add code to register the credentials:

```js
// Registration
async function register(optionsJSON) {

>   const options = convertOptions(JSON.parse(optionsJSON));

Reviewer: Drop the `convertOptions` helper and inline it here:

```suggestion
  const options = JSON.parse(optionsJSON);
  options.challenge = Uint8Array.fromBase64(options.challenge);
  options.user.id = Uint8Array.fromBase64(options.user.id);
```

  const credential = await navigator.credentials.create({ publicKey: options });
  console.log("Credential created:", credential);
  return credential;
}
```

The JSON you use to perform the registration, with `rp.id` set to an external domain, looks something like this:

```json
{
  "rp": {
    "name": "Example Corp",
    "id": "example.com"
  },
  "user": {
    "id": "dXNlcjEyMw==",
    "name": "user@example.com",
    "displayName": "Example User"
  },
  "challenge": "cmFuZG9tQ2hhbGxlbmdl",
  "pubKeyCredParams": [{ "type": "public-key", "alg": -7 }]
}
```

### Authenticate the credential

> To authenticate the credentials, call `navigator.credentials.get()` with the `rpId` in the JSON set to the target domain. Again, you need to convert binary fields, such as `challenge` and `user.id`, to `ArrayBuffer` instances before asserting:

Reviewer: ```suggestion
To authenticate the credentials, call `navigator.credentials.get()` with the `rpId` in the JSON set to the target domain. Again, you need to convert binary fields, such as `challenge`, from the base64-encoded string as used in the example, to `Uint8Array` instances before asserting:
```

```js
async function authenticate(optionsJSON) {

>   const options = convertOptions(JSON.parse(optionsJSON));

Reviewer: ```suggestion
  const options = JSON.parse(optionsJSON);
  options.id = Uint8Array.fromBase64(options.challenge);
  if (Array.isArray(options.allowCredentials)) {
    for (const ac of options.allowCredential) {
      ac.id = Uint8Array.fromBase64(ac.id);
    }
  }
```
